### PR TITLE
SDK v2 development

### PR DESCRIPTION
**Pull request by fatihkurtoglu**
Python SDK v2.0 development completed.

More details are available in the migration guide and will be provided in Release Notes. 
You can refer to the internal PRD as well.

High level summary:

- Object model updates:
  - Task/Batch/Project models updated
  - A new method `as_dict()` introduced to access object as a dict 

- New ways to retrieve the list of tasks/batches:
  - Users don't need to handle pagination and tokens/offsets/timeouts anymore
  - `tasks_list` and `batches_list` are the new generator methods for bulk retrieval
  - Deprecated some methods to avoid naming confusion

- API:
  - Isolated API access into a different class
  - Enabled HTTP retry for certain error codes
  - Improved error handling by differentiating exception types
 
- Infra improvements:
  - Enabled type hinting across the package
  - New code standards applied via Pylint, flake8 and black
  - Integrated `pre-commit` for a better/consistent developer experience
  - `publish.sh` introduced for an automated publish to PyPI
  - New `pytest` test cases are added
  - Updated setup.py for v2

- Documentation
  - New **Migration guide** for v2
  - New **Developer Guide** (how to setup repo env and configure pre-commit)
  - Updated deployment and publishing guide
  - Updated README for v2
  - Made README to be available in PyPI

Testing results with pytest:

```
======================================== test session starts =========================================
platform darwin -- Python 3.6.12, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /Users/fatihkurtoglu/git/scaleapi-python-client
collected 31 items                                                                                                                                                                                   

tests/test_client.py ..............................                                             [100%]
======================================== 31 passed in 18.31s =========================================
```

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -61,46 +78,49 @@ __ https://docs.scale.com/reference#general-image-annotation
...
+    print(task.status)  # Task status ('pending', 'completed', 'error', 'canceled')
+    print(task.response) # If task is complete
 
 List Tasks
 ^^^^^^^^^^
 
-Retrieve a list of tasks, with optional filter by start and end date/time. Paginated with `next_token`. The return value is a `scaleapi.Tasklist`, which acts as a list, but also has fields for the total number of tasks, the limit and offset, and whether or not there's more. Check out `Scale's API documentation`__ for more information.
+Retrieve a list of `Task` objects, with optional filters for: `project_name, batch_name, type, status, review_status, unique_id, completed_after, completed_before, updated_after, updated_before, created_after, created_before` and `tags`.
+
+This method is a generator and yields tasks. It can be wrapped in a `list` statement if a Task list is needed.
```
Comment (shaun-scale) [inline README.rst:99]: Can we add a few more sentences here about how a generator works / the implications. A lot of less experienced developers will also be using this method so we'll want to give at least a little bit of background about how to take advantage of this.

File: README.rst [hunk 2]
```diff
@@ -23,11 +37,11 @@ Tasks
 _____
 
 Most of these methods will return a `scaleapi.Task` object, which will contain information
-about the json response (task_id, status, etc.).
+about the json response (task_id, status, params, response etc.).
```
Comment (shaun-scale) [inline README.rst:40]: Smallest nitpick in the world, add a comma between response and etc. `response, etc.`

File: README.rst [hunk 3]
```diff
@@ -232,7 +259,7 @@ __ https://docs.scale.com/reference#batch-list
     projects = client.projects()
     for project in projects:
         counter += 1
-        print('Downloading project %s | %s | %s' % (counter, project['name'], project['type']))
+        print(f'Downloading project {counter} | {project.name} | { project.type}')
```
Comment (shaun-scale) [inline README.rst:262]: Extraneous space I think in the last param here `{ project.type}`

File: README.rst [hunk 4]
```diff
@@ -245,23 +272,38 @@ __ https://docs.scale.com/reference#project-update-parameters
 
     data = client.update_project(
         project_name='test_project',
-        pathc = false,
+        patch = false,
```
Comment (shaun-scale) [inline README.rst:275]: oof - thank you!

File: README.rst [hunk 5]
```diff
@@ -245,23 +272,38 @@ __ https://docs.scale.com/reference#project-update-parameters
...
-    except scaleapi.ValidationError as e:
-        print(e.code)  # 400
-        print(e.message)  # missing param X
+    from scaleapi.exceptions import ScaleException
+
+    try:
+        client.create_task(TaskType.TextCollection, attachment='Some parameters are missing.')
+    except ScaleException as err:
+        print(err.code)  # 400
+        print(err.message)  # Parameters is invalid, reason: "attachments" is required
```
Comment (shaun-scale) [inline README.rst:306]: Is this the actual error message we provide? 😅 It should be "Parameters **are** invalid" or "The `parameters` field is invalid" I think. If this is exactly what we do return though, it's probably fine to leave as is.

**Review by shaun-scale — APPROVED** [review]
Left small comments - only big comment, and maybe it's too late.. I think I prefer `all_tasks` to `tasks_all` but if it is too late... (I think it might be.. 😅 ) - I'm fine.

**Comment by fatihkurtoglu** [comment]
> Left small comments - only big comment, and maybe it's too late.. I think I prefer `all_tasks` to `tasks_all` but if it is too late... (I think it might be.. 😅 ) - I'm fine.

@shaun-scale 
No, it's not too late. Actually, there is a bit gray area on how to name those methods (get/fetch/list/retrieve/$s). 

Existing methods were named as `tasks` and `projects` (as named per endpoint). I also renamed `list_batches` to `batches` to be consistent with the existing naming.

For naming new yield/generator methods, some prefer `_yield` or `_gen` suffixes, making sense for differentiation but more developer-oriented. I'm open to discussing that, and it's not a big change either. 

We can use `all` as a pre-fix as well, and this is better for grammar! I was thinking of an auto-complete scenario where IDE provides the other method when a dev types `client.task`

Another alternative is to use `get_tasks()` and `get_batches()` since we already have singular versions to retrieve one, and plural ones can serve to retrieve the list.

File: README.rst (hunk 5 above)
Comment (fatihkurtoglu) [reply README.rst:306]: Unfortunately, this is the actual message. I'll send a PR to fix it to `Parameter is` :)

File: README.rst (hunk 2 above)
Comment (fatihkurtoglu) [reply README.rst:40]: Should've run grammarly on readme as well :) Thanks for the catch!

**Comment by shaun-scale** [comment]
I think I like the plural versions (`get_tasks()` and `get_batches()`, etc.) - seems very readable and clear, plus would help with the auto-completion :)

**Pull request merged**